scipy/signal/spectral.py

```diff
@@ -267,121 +267,495 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
+        Length of the FFT used, if a zero padded FFT is desired.  If None,
+        the FFT length is `nperseg`. Defaults to None.
+    detrend : str or function or False, optional
+        Specifies how to detrend each segment. If `detrend` is a string,
+        it is passed as the ``type`` argument to `detrend`.  If it is a
+        function, it takes a segment and returns a detrended segment.
+        If `detrend` is False, no detrending is done.  Defaults to 'constant'.
+    return_onesided : bool, optional
+        If True, return a one-sided spectrum for real data. If False return
+        a two-sided spectrum. Note that for complex data, a two-sided
+        spectrum is always returned.
+    scaling : { 'density', 'spectrum' }, optional
+        Selects between computing the power spectral density ('density')
+        where Pxx has units of V**2/Hz if x is measured in V and computing
+        the power spectrum ('spectrum') where Pxx has units of V**2 if x is
+        measured in V. Defaults to 'density'.
+    axis : int, optional
+        Axis along which the periodogram is computed; the default is over
+        the last axis (i.e. ``axis=-1``).
+    mode : str, optional
+        Defines what kind of return values are expected. Options are ['psd',
+        'complex', 'magnitude', 'angle', 'phase'].
+
+    Returns
+    -------
+    freqs : ndarray
+        Array of sample frequencies.
+    result : ndarray
+        Array of output data, contents dependant on *mode* kwarg.
+    t : ndarray
+        Array of times corresponding to each data segment
+
+    References
+    ----------
+    stackoverflow: Rolling window for 1D arrays in Numpy?
+    <http://stackoverflow.com/a/6811241>
+    stackoverflow: Using strides for an efficient moving average filter
+    <http://stackoverflow.com/a/4947453>
+
+    Notes
+    -----
+    Adapted from matplotlib.mlab
+    '''
+    if mode not in ['psd', 'complex', 'magnitude', 'angle', 'phase']:
+        raise ValueError("Unknown value for mode %s, must be one of: "
+                         "'default', 'psd', 'complex', "
+                         "'magnitude', 'angle', 'phase'" % mode)
+
+    # If x and y are the same object we can save ourselves some computation.
+    same_data = y is x
+
+    if not same_data and mode != 'psd':
+        raise ValueError("x and y must be equal if mode is not 'psd'")
+
+    # Ensure we have np.arrays, get outdtype
     x = np.asarray(x)
+    if not same_data:
+        y = np.asarray(y)
+        outdtype = np.result_type(x,y,np.complex64)
+    else:
+        outdtype = np.result_type(x,np.complex64)
+
+    if x.ndim > 1:
+        if axis != -1:
+            x = np.rollaxis(x, axis, len(x.shape))
+            if not same_data and y.ndim > 1:
+                y = np.rollaxis(y, axis, len(y.shape))
+            # Output is going to have new last axis for window index
+            if axis < 0:
+                axis = len(np.broadcast(x,y).shape)-axis
 
     if x.size == 0:
```

Comment:
This check should come earlier.  Consider:
```
In [14]: a = np.array([[[],[],[]],[[],[],[]]])

In [15]: a.shape
Out[15]: (2, 3, 0)

In [16]: signal.welch(a, axis=1)
Out[16]:
(array([], shape=(2, 0, 3), dtype=float64),
 array([], shape=(2, 0, 3), dtype=float64))
```
Bit of corner case, I suppose.  Maybe this should be a test?  The currently shipping version of `welch` gets this right.